### PR TITLE
create artifacts and dependencies

### PR DESCRIPTION
**Pull request by ilaner**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/create_artifacts/create_artifacts_v2.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    PackDependencies
+
+
+class ContentArtifactManager:
+
+    def __init__(self,
+                 marketplace: MarketplaceVersions,
+                 output: Path,
+                 zip: bool,
+                 dependencies: bool,
```
Comment (DeanArbel) [inline demisto_sdk/commands/create_artifacts/create_artifacts_v2.py:20]: since it's a bool, please put False as default

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/nodes.py [hunk 2]
```diff
@@ -77,38 +77,47 @@ def create_nodes_by_type(
...
     RETURN p AS pack, collect(c) AS content_items
     """
-    return run_query(tx, query).data()
-
-
-def get_all_integrations_with_commands(
+    packs: List[Pack] = []
+    integrations_to_commands = _get_all_integrations_with_commands(tx)
+    for item in run_query(tx, query).data():
+        pack = item.get('pack')
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/nodes.py:94]: First run the original query, and only when it succeeds run the `integrations_to_command` query.

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 3]
```diff
@@ -1,14 +1,44 @@
...
 
 from demisto_sdk.commands.common.constants import MarketplaceVersions
 from demisto_sdk.commands.content_graph.common import ContentType
 
+content_type_to_model: Dict[ContentType, Type['BaseContent']] = {}
 
-class BaseContent(ABC, BaseModel):
+
+class ContentModelMetaclass(ModelMetaclass):
+    def __new__(cls, name, bases, namespace, content_type: ContentType = None, **kwargs):
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/objects/base_content.py:15]: Please check whether it's possible to implement the same functionality without explicitly providing `content_type` (e.g. getting it from the class name)

File: demisto_sdk/commands/content_graph/objects/integration_script.py [hunk 4]
```diff
@@ -1,8 +1,26 @@
...
 class IntegrationScript(ContentItem):
     type: str
     docker_image: Optional[str]
+    description: Optional[str]
+
+    def dump(self, dir: Path, marketplace: MarketplaceVersions) -> None:
+        # demisto-sdk unify self.path -> path
+        dir.mkdir(exist_ok=True, parents=True)
+        try:
+            IntegrationScriptUnifier(input=str(self.path.parent), output=str(dir), marketplace=marketplace).unify()
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/objects/integration_script.py:22]: Is there a way to check whether it's unified already?

File: demisto_sdk/commands/content_graph/objects/content_item.py [hunk 5]
```diff
@@ -14,3 +19,25 @@ class ContentItem(BaseContent):
...
     deprecated: bool
     description: Optional[str]
+
+    def summary(self) -> dict:
+        return self.dict(include=self.included_in_metadata(), by_alias=True)
+
+    def included_in_metadata(self) -> Set[str]:
+        raise NotImplementedError('Should be implemented in subclasses')
+
+    def normalize_file_name(self, name: str) -> str:
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/objects/content_item.py:29]: Please a docstring explaining why this is needed. This seems like a workaround.

File: demisto_sdk/commands/content_graph/objects/content_item.py [hunk 6]
```diff
@@ -14,3 +19,25 @@ class ContentItem(BaseContent):
...
+        for prefix in ContentType.prefixes():
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/objects/content_item.py:30]: Please add a comment explaining why this is in a for loop (with no break)

Comment (DeanArbel) [reply demisto_sdk/commands/content_graph/objects/content_item.py:30]: please rename to server_names

File: demisto_sdk/commands/content_graph/objects/content_item.py [hunk 7]
```diff
@@ -14,3 +19,25 @@ class ContentItem(BaseContent):
...
+            name = name.replace(f'{prefix}-', '')
+
+        return f'{self.content_type.server_name}-{name}'
+
+    def dump(self, dir: DirectoryPath, _: MarketplaceVersions) -> None:
+        dir.mkdir(exist_ok=True, parents=True)
+        shutil.copy(self.path, dir / self.normalize_file_name(self.path.name))
+
+    def to_id_set_entity(self) -> dict:
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/objects/content_item.py:39]: Please add a short docstring (mention this is a temp solution)

File: demisto_sdk/commands/common/content/objects/pack_objects/readme/readme.py [hunk 8]
```diff
@@ -42,7 +42,7 @@ def handle_marketplace_tags(self):
         try:
             with open(self._path, 'r+') as f:
                 text = f.read()
-                parsed_text = get_mp_tag_parser().parse_text(text)
+                parsed_text = get_mp_tag_parser(marketplace=None).parse_text(text)
```
Comment (DeanArbel) [inline demisto_sdk/commands/common/content/objects/pack_objects/readme/readme.py:45]: This seems redundant

File: demisto_sdk/commands/content_graph/interface/graph.py [hunk 9]
```diff
@@ -23,39 +30,55 @@ def validate_graph(self) -> None:
...
+    @abstractmethod
+    def get_first_level_dependencies(self, marketplace: MarketplaceVersions) -> Dict[str, Dict[str, Any]]:
+        pass
+
+    @abstractmethod
+    def create_pack_dependencies(self):
+        pass
+
+    @abstractmethod
+    def get_connected_nodes_by_relationship_type(
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/interface/graph.py:73]: Please add a docstring (make sure to explain the input/output).

File: demisto_sdk/commands/content_graph/content_graph_builder.py [hunk 10]
```diff
@@ -61,7 +61,7 @@ def create_graph(self) -> None:
         self.content_graph.create_indexes_and_constraints()
         self.content_graph.create_nodes(self.nodes)
         self.content_graph.create_relationships(self.relationships)
-        self.content_graph.validate_graph()
+        # self.content_graph.validate_graph()
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/content_graph_builder.py:64]: Please add this back - if there are specific validations issues handle them specifically

File: demisto_sdk/commands/content_graph/interface/graph.py [hunk 11]
```diff
@@ -23,39 +30,55 @@ def validate_graph(self) -> None:
...
         self,
+        marketplace: MarketplaceVersions,
         content_type: Optional[ContentType] = None,
         **properties,
-    ) -> Any:
+    ) -> List[BaseContent]:
         pass
 
     @abstractmethod
     def get_single_node(
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/interface/graph.py:52]: Why is this needed?

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 12]
```diff
@@ -21,3 +51,7 @@ class Config:
 
     def to_dict(self) -> Dict[str, Any]:
         return json.loads(self.json())
+
+    @abstractmethod
+    def dump(self, path: DirectoryPath, marketplace: MarketplaceVersions) -> None:
```
Comment (DeanArbel) [inline demisto_sdk/commands/content_graph/objects/base_content.py:56]: Is `marketplace` needed? It should be set in the os.env.

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 13]
```diff
@@ -113,11 +121,63 @@ class PackMetadata(BaseModel):
...
     contributors: Optional[List[str]] = None
     content_items: PackContentItems = Field(alias='contentItems', exclude=True)
     relationships: Relationships = Field(Relationships(), exclude=True)
 
+    def dump_metadata(self, path: Path) -> None:
+        metadata = self.dict(exclude={'path', 'node_id', 'content_type'})
+        metadata['contentItems'] = {}
+        for content_item in self.content_items:
+            if content_item.content_type == ContentType.TEST_PLAYBOOK:
+                continue
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:135]: redundant due to the following try-except block

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 14]
```diff
@@ -113,11 +121,63 @@ class PackMetadata(BaseModel):
...
 
-class Pack(BaseContent, PackMetadata):
+class Pack(BaseContent, PackMetadata, content_type=ContentType.PACK):
     path: Path
     contributors: Optional[List[str]] = None
     content_items: PackContentItems = Field(alias='contentItems', exclude=True)
     relationships: Relationships = Field(Relationships(), exclude=True)
 
+    def dump_metadata(self, path: Path) -> None:
+        metadata = self.dict(exclude={'path', 'node_id', 'content_type'})
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:131]: is it possible to explicitly include `contentItems` instead of the logic below?

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 15]
```diff
@@ -113,11 +121,63 @@ class PackMetadata(BaseModel):
...
+        with open(path, 'r+') as f:
+            try:
+                text = f.read()
+                parsed_text = get_mp_tag_parser(marketplace).parse_text(text)
+                if len(text) != len(parsed_text):
+                    f.seek(0)
+                    f.write(parsed_text)
+                    f.truncate()
+            except Exception as e:
+                logger.error(f'Failed dumping readme: {e}')
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:159]: let's try to make this file with as less non-relevant logic in pack object as possible
for example we can move this to a classmethod of MarketplaceParser

**Review by dantavori — CHANGES_REQUESTED** [review]
amazing work! see notes

File: demisto_sdk/commands/content_graph/objects/integration.py [hunk 16]
```diff
@@ -1,23 +1,49 @@
...
 
-from demisto_sdk.commands.content_graph.objects.integration_script import \
-    IntegrationScript
+from demisto_sdk.commands.content_graph.common import ContentType, RelationshipType
+from demisto_sdk.commands.content_graph.objects.integration_script import IntegrationScript
 
 
-class Command(BaseModel):
+class BaseCommand(BaseModel):
     name: str
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/integration.py:11]: no need to pass element id anymore

File: demisto_sdk/commands/content_graph/objects/integration.py [hunk 17]
```diff
@@ -1,23 +1,49 @@
...
+                name=r.related_to.name,  # type: ignore[union-attr]
+                deprecated=r.deprecated,
+                description=r.description,
+            )
+            for r in self.relationshipss
+            if not r.is_nested and r.relationship_type == RelationshipType.HAS_COMMAND
+        ]
+
+    def included_in_metadata(self):
+        return {"name", "description", "category", "commands"}
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/integration.py:44]: no commands anymore

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 18]
```diff
@@ -1,117 +1,261 @@
...
+                Neo4jContentGraphInterface._id_to_obj[element_id] = server_content
+
+            elif content_type == ContentType.COMMAND:
+                command = BaseCommand.parse_obj(node)
+                Neo4jContentGraphInterface._id_to_obj[element_id] = command
+
+            else:
+                model = content_type_to_model.get(content_type)
+                if not model:
+                    # this currently happens for command node?
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:93]: remove comment

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 19]
```diff
@@ -31,14 +35,14 @@ class Neo4jContentGraphInterface(ContentGraphInterface):
     def __init__(
```
[the remark this answers is not in the document]
Comment (ilaner) [reply demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:46]: wdym?

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 20]
```diff
@@ -1,117 +1,237 @@
...
+    ) -> List[BaseContent]:
+        """
+        match('xsoar', content_type=INTEGRATION)
+        """
         with self.driver.session() as session:
-            return session.read_transaction(search_nodes, content_type, **properties)
-
-    def get_single_node(
+            result: List[
+                Tuple[graph.Node, List[graph.Relationship], List[graph.Node]]
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:134]: maybe return named tuple from neo4j interface

File: demisto_sdk/commands/content_graph/objects/integration.py [hunk 21]
```diff
@@ -1,23 +1,70 @@
...
+        return values["name"]
 
     class Config:
         orm_mode = True
 
 
-class Integration(IntegrationScript):
+class Command(BaseCommand):
+    name: str
+    deprecated: bool = False
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/integration.py:33]: check if can use this class

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 22]
```diff
@@ -1,117 +1,237 @@
...
+        nodes_to: List[graph.Node],
+    ) -> None:
+        # Command cannot be the start node of relationship
+        rel = {
+            RelationshipData(
+                relationship_type=rel.type,
+                source=Neo4jContentGraphInterface._id_to_obj[rel.start_node.element_id],
+                target=Neo4jContentGraphInterface._id_to_obj[rel.end_node.element_id],
+                related_to=Neo4jContentGraphInterface._id_to_obj[node_to.element_id],
+                is_nested=rel.start_node.element_id != node_from.element_id and
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:187]: reverse condition

File: demisto_sdk/commands/content_graph/objects/integration_script.py [hunk 23]
```diff
@@ -1,8 +1,38 @@
...
+
+yaml = YAML_Handler()
+
+logger = logging.getLogger("demisto-sdk")
 
 
 class IntegrationScript(ContentItem):
     type: str
     docker_image: Optional[str]
+    description: Optional[str]
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/integration_script.py:18]: check if works

File: demisto_sdk/commands/content_graph/tests/create_content_graph_test.py [hunk 24]
```diff
@@ -223,104 +285,115 @@ def test_create_content_graph_single_pack(
...
+            assert len(returned_pack.content_items.script) == 1
+
+            returned_integration = returned_pack.content_items.integration[0]
+            assert returned_integration.commands[0].name == "test-command"
+            uses_integration = returned_integration.uses[0]
+            assert uses_integration.object_id == "SampleClassifier"
+            tested_by = returned_integration.tested_by[0]
+            assert tested_by.object_id == "SampleTestPlaybook"
+
+            imports = returned_integration.imports
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/tests/create_content_graph_test.py:391]: dont assume only one import

File: demisto_sdk/commands/content_graph/objects/base_content.py (hunk 12 above)
Comment (ilaner) [reply demisto_sdk/commands/content_graph/objects/base_content.py:98]: let's keep it we would like in the future to run one script for both marketplaces

File: demisto_sdk/commands/content_graph/objects/relationship.py [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+@dataclass
+class RelationshipData:
+    relationship_type: RelationshipType
+    source: Union[BaseContent, Command]
+    target: Union[BaseContent, Command]
+
+    related_to: Union[BaseContent, Command]
+    is_direct: bool = True
+
+    mandatorily: bool = False
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/relationship.py:23]:
```suggestion

    # USES relationship properties
    mandatorily: bool = False
```

File: demisto_sdk/commands/content_graph/objects/relationship.py [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+    description: Optional[str] = None
+    deprecated: bool = False
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/relationship.py:25]:
```suggestion

    # HAS_COMMAND relationship properties
    description: Optional[str] = None
    deprecated: bool = False
```

File: demisto_sdk/commands/content_graph/interface/graph.py [hunk 27]
```diff
@@ -15,48 +21,56 @@ def create_nodes(self, nodes: Dict[ContentType, List[Dict[str, Any]]]) -> None:
...
+        packs = self.search(
+            marketplace, content_type=ContentType.PACK
+        )
+        return ContentDTO(packs=packs)
 
     @abstractmethod
-    def get_relationships_by_type(self, relationship: Relationship) -> Any:
-        pass
+    def create_pack_dependencies(self):
+        ContentGraphInterface._with_dependencies = True
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/graph.py]: why is this needed?

**Review by dantavori — COMMENTED** [review]
looks good,
I'm not in favor of the auto-formatting (especially the change of `'` to `"`, our convention has always been `'`) - it caused too much diff for nothing really.

File: demisto_sdk/commands/content_graph/interface/graph.py (hunk 27 above)
Comment (ilaner) [reply demisto_sdk/commands/content_graph/interface/graph.py:73]: If the graph is already has dependencies, we don't wan't to execute this query again

File: demisto_sdk/commands/common/tools.py [hunk 28]
```diff
@@ -190,8 +190,10 @@ def get_log_verbose() -> bool:
     return LOG_VERBOSE
 
 
-def get_mp_tag_parser():
+def get_mp_tag_parser(marketplace: Optional[MarketplaceVersions] = None):
     global MARKETPLACE_TAG_PARSER
+    if marketplace:
+        return marketplace
```
Comment (DeanArbel) [inline demisto_sdk/commands/common/tools.py:196]: Please remove this change.
This behavior is only being used by one function and makes little sense.

**Review by DeanArbel — APPROVED** [review]
Please refrain from running `black` on existing files that were not formatted with black.

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 29]
```diff
@@ -1,23 +1,108 @@
...
+            if r.relationship_type == RelationshipType.USES and r.related_to == r.target
+        ]
+
+    @property
+    def tested_by(self) -> List["TestPlaybook"]:
+        return [
+            r.related_to
+            for r in self.relationships_data
+            if r.relationship_type == RelationshipType.TESTED_BY and r.related_to == r.target
+        ]
```
[the remark this answers is not in the document]
Comment (ilaner) [reply demisto_sdk/commands/content_graph/objects/base_content.py]: moved to content item

File: demisto_sdk/commands/content_graph/objects/pack.py (hunk 14 above)
Comment (ilaner) [reply demisto_sdk/commands/content_graph/objects/pack.py]: Tried it, but it didn't work. It includes only the `contentItems`

File: demisto_sdk/commands/content_graph/interface/graph.py (hunk 27 above)
Comment (dantavori) [reply demisto_sdk/commands/content_graph/interface/graph.py:83]: let's talk about this too

**Review by dantavori — APPROVED** [review]
